src/core/service.c

```diff
@@ -1401,6 +1401,31 @@ static bool service_exec_needs_notify_socket(Service *s, ExecFlags flags) {
         return s->notify_access != NOTIFY_NONE;
 }
 
+static Service *service_get_on_failure_of_dep(Service *s) {
+        assert(s);
+
+        /* If this service had a single reverse OnFailure= dependency of type
+         * UNIT_SERVICE which has failed, return that dependency. */
+
+        Unit *u = UNIT(s);
+        Unit *trigger;
+        Hashmap *deps;
+        void *v;
+
+        deps = unit_get_dependencies(u, UNIT_ON_FAILURE_OF);
```

Comment:
hmm, please don't enumerate things this way.

please use the unit dependency "atoms", i.e. define a new flag in UnitDependencyAtom called UNIT_ATOM_PICK_UP_EXIT_STATUS or so. Then in the table in atom_map[] make sure that both UNIT_ON_FAILURE_OF and UNIT_ON_SUCCESS_OF dep types include this atom. 

With that in place you can then easily enumerate the deps:

UNIT_FOREACH_DEPENDENCY(other, UNIT(s), UNIT_ATOM_PICK_UP_EXIT_STATUS) {
   …
}

Reply:
explanation for the atom stuff: many of our high-level dep types share the same properties. like here for example: both OnSuccess= and OnFailure= want this propagation of exit status stuff. Hence we decided to change deps to be made up of so called "atoms" that indicate what a specific high-level dep type is actually made up of, and then always act on the atoms, never the deps themselves.